pkg/client/cache/delta_fifo.go

```diff
@@ -43,13 +45,13 @@ import (
 //       affects error retrying.
 //
 // Also see the comment on DeltaFIFO.
-func NewDeltaFIFO(keyFunc KeyFunc, compressor DeltaCompressor, knownObjectKeys KeyLister) *DeltaFIFO {
+func NewDeltaFIFO(keyFunc KeyFunc, compressor DeltaCompressor, knownObjects KeyLookup) *DeltaFIFO {
```

Comment:
As a concrete suggestion, why not leave this the same, but add the interface below, too, and say in the docs that if knownObjectKeys supports it, it will be called instead of just dropping in finalStateUnknown markers.

```
type KeyGetter interface {
  GetByKey(key string) (interface{}, bool, error)
}
```

The problem is that the known objects may not have the *final* state, so it may be just plain wrong to pass the last known state as the final state in a deletion.